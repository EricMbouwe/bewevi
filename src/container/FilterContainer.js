import { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import Filter from '../component/Filter';
import {
  fetchCategoryName, fetchFilterName, updateFilterPokemon,
} from '../redux';

const FilterContainer = () => {
  const categoryNameList = [
    ['Habitat', 'pokemon-habitat/'],
    ['Shape', 'pokemon-shape/'],
    ['Color', 'pokemon-color/'],
  ];

  const categoryList = useSelector(state => state.pokemon.filter.categoryList);
  const [category, setCategory] = useState('Filter By Category');
  const [filterName, setFilterName] = useState('Filter Name');

  const dispatch = useDispatch();

  const handleChangeCategory = e => {
    setCategory(e.target.value);
    if (e.target.value !== 'All') {
      dispatch(fetchCategoryName(e.target.value));
    } else {
      setFilterName('Filter Name');
      dispatch(updateFilterPokemon([]));
    }
  };

  const handleFilterName = e => {
    setFilterName(e.target.value);
    if (e.target.value !== 'All') {
      dispatch(fetchFilterName(`${category}${e.target.value}`));
    } else {
      dispatch(updateFilterPokemon([]));
    }
  };

  return (
    <Filter
      handleChangeCategory={handleChangeCategory}
      handleFilterName={handleFilterName}
      category={category}
      filterName={filterName}
      categoryNameList={categoryNameList}
      categoryList={categoryList}
    />
  );
};

export default FilterContainer;
