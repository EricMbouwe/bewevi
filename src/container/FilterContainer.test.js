import { render, screen } from '@testing-library/react';
import React from 'react';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';

import FilterContainer from './FilterContainer';

const initStore = { plant: { filterInput: '' } };
const mockStore = configureStore();
const store = mockStore(initStore);
store.dispatch = jest.fn();

let renderReadyComponent;

let setState;
let useStateSpy;

beforeEach(() => {
  setState = jest.fn();
  useStateSpy = jest.spyOn(React, 'useState');
  useStateSpy.mockImplementation(init => [init, setState]);

  renderReadyComponent = (
    <Provider store={store}>
      <FilterContainer />
    </Provider>
  );
});

describe('<FilterContainer />', () => {
  it('is triggering handleClick when the button is clicked', () => {
    render(
      renderReadyComponent,
    );
    expect(store.dispatch.mock.calls.length).toEqual(0);
    userEvent.click(screen.getByText('Filter Results'));
    expect(store.dispatch.mock.calls.length).toEqual(1);
  });

  it('is triggering handleChangeFilter when the filter is changed', () => {
    render(
      renderReadyComponent,
    );
    userEvent.selectOptions(screen.getByDisplayValue('Common Name'), ['Year']);
    expect(setState.mock.calls).toEqual([['year']]);
  });

  it('is triggering handleChangeInput when input is changed', () => {
    render(
      renderReadyComponent,
    );
    userEvent.type(screen.getByPlaceholderText('Filter'), 'I');
    expect(setState.mock.calls).toEqual([['I']]);
  });

  it('renders correctly', () => {
    const renderedContainer = render(renderReadyComponent);
    expect(renderedContainer).toMatchSnapshot();
  });
});
