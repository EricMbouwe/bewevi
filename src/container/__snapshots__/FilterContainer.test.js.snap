// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilterContainer /> renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="filter main"
      >
        <select
          class="filter filter-select"
          name="filter"
        >
          <option
            value="common_name"
          >
            Common Name
          </option>
          <option
            value="year"
          >
            Year
          </option>
          <option
            value="scientific_name"
          >
            Scientific Name
          </option>
          <option
            value="family_common_name"
          >
            Common Family Name
          </option>
          <option
            value="family"
          >
            Family Name
          </option>
        </select>
        <input
          class="filter filter-input"
          placeholder="Filter"
          type="text"
          value=""
        />
        <button
          class="filter filter-button"
          type="button"
        >
          Filter Results
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="filter main"
    >
      <select
        class="filter filter-select"
        name="filter"
      >
        <option
          value="common_name"
        >
          Common Name
        </option>
        <option
          value="year"
        >
          Year
        </option>
        <option
          value="scientific_name"
        >
          Scientific Name
        </option>
        <option
          value="family_common_name"
        >
          Common Family Name
        </option>
        <option
          value="family"
        >
          Family Name
        </option>
      </select>
      <input
        class="filter filter-input"
        placeholder="Filter"
        type="text"
        value=""
      />
      <button
        class="filter filter-button"
        type="button"
      >
        Filter Results
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
